### ISSUE TITLE
Allow automatic use of Headscale certs?

### ISSUE THREAD

**Issue by edgar-vincent** [caddyserver/caddy#5197]
Hello,

The [docs](https://caddyserver.com/docs/caddyfile/directives/tls#tailscale) mention that Headscale, a libre implementation of the Headscale control server, can be used with Caddy. However, Caddy only automatically uses Tailscale for `*.ts.net` domains, which are specific to Tailscale services. Headscale users, on the other hand, obviously get to define their own domain name.

Would it be possible, for example, to be able to define the domain used by Headscale in Caddy's configuration, in order for it to use its TLS certificates automatically?

Thanks a lot!

EV

**Comment by mholt** [comment]
Yeah, I think we can do this. Not sure how yet, but most likely doable!

**Comment by edgar-vincent** [comment]
Thank you very much!

**Comment by motiejus** [comment]
This would require coupling of the user's dns server and headscale or tailscaled. HS or TSd would need to know how to nudge the DNS server to add the dns txt records for verification. This is hard, because everyone uses a different DNS server and API to change the records.

It works so easily with tailscale, because tailscale the company control `*.ts.net`, tailscale service and tailscaled.

Caddy asking for a cert is the icing. The hard part is getting the pieces to talk to each other.

I have implemented https for headscale domains myself, here is a public domain `bitwarden.jakstys.lt` pointing to `100.<...>` ip: https://git.jakstys.lt/motiejus/config/src/commit/d60677822f3e340c1f6a96533de432884e10bc77/data.nix#L170

Implementation of dns txt verification is in the same repo: https://git.jakstys.lt/motiejus/config/src/commit/d60677822f3e340c1f6a96533de432884e10bc77/modules/services/nsd-acme/default.nix

As you can see, it's convoluted.

**Comment by francislavoie** [comment]
@motiejus I don't use Tailscale nor Headscale so I'm not equipped to help much here, but it sounds like we need to integrate Caddy's DNS plugins to make this happen. I don't know what the triggers are here, but that does sound solvable.

**Comment by teleclimber** [comment]
Note that Headscale does not yet fetch certs to use with server over TLS but it probably will some day:

https://github.com/juanfont/headscale/issues/1921
